source/MRMesh/MRToolPath.cpp

```diff
@@ -374,5 +559,123 @@ std::string exportToolPathToGCode( const std::vector<GCommand>& commands )
 
     return gcode.str();
 }
+
+float distSqrToLineSegment( const MR::Vector2f p, const MR::Vector2f& seg0, const MR::Vector2f& seg1 )
+{
+    const auto segDir = seg1 - seg0;
+    const auto len2 = segDir.lengthSq();
+    if ( len2 < 1e-30 )
+    {
+        return ( seg0 - p ).lengthSq();
+    }
+
+    const auto tmp = cross( p - seg0, segDir );
+    return ( tmp * tmp ) / len2;
+}
+
+
+std::vector<GCommand> replaceStraightSegmentsWithOneLine( const std::span<GCommand>& path, float eps, float maxLength, Axis axis )
+{
+    if ( path.size() < 3 )
+        return {};
+
+    std::vector<GCommand> res;
+
+    const float epsSq = eps * eps;
+    const float maxLengthSq = maxLength * maxLength;
+    int startIdx = 0, endIdx = 0;
+    for ( int i = startIdx + 1; i < path.size(); ++i )
+    {
+        const auto& d0 = path[startIdx];
+        const auto& d2 = path[i];
+
+        const Vector2f p0 = d0.project( axis );
+        Vector2f p2 = d2.project( axis );
+
+        bool canInterpolate = ( p0 - p2 ).lengthSq() < maxLengthSq // don't merge too long lines
+            && d2.type == MoveType::Linear; // don't merge arcs
```

Comment:
should we check length in 3d?

Reply:
Here we are sure that all points in the segment lay in the same plane, so it's enough to compute 2D length